pydropsonde/circles.py

```diff
@@ -15,6 +20,179 @@ class Circle:
+            dx=(["sonde_id", "alt"], delta_x.values, delta_x_attrs),
+            dy=(["sonde_id", "alt"], delta_y.values, delta_y_attrs),
+        )
+
+        self.circle_ds = self.circle_ds.assign(new_vars)
+
+        return self
+
+    @staticmethod
+    def fit2d(x, y, u):
+        a = np.stack([np.ones_like(x), x, y], axis=-1)
+
+        invalid = np.isnan(u) | np.isnan(x) | np.isnan(y)
+        u_cal = np.where(invalid, 0, u)
+        a[invalid] = 0
+
+        a_inv = np.linalg.pinv(a)
+        intercept, dudx, dudy = np.einsum("...rm,...m->r...", a_inv, u_cal)
+
+        return intercept, dudx, dudy
+
+    def fit2d_xr(self, x, y, u, sonde_dim="sonde_id"):
+        return xr.apply_ufunc(
+            self.__class__.fit2d,  # Call the static method without passing `self`
+            x,
+            y,
+            u,
+            input_core_dims=[
+                [sonde_dim],
+                [sonde_dim],
+                [sonde_dim],
+            ],  # Specify input dims
+            output_core_dims=[(), (), ()],  # Output dimensions as scalars
+        )
+
+    def apply_fit2d(self):
+        variables = ["u", "v", "q", "ta", "p"]
+
+        description_names = [
+            "eastward wind",
+            "northward wind",
+            "specific humidity",
+            "air temperature",
+            "air pressure",
+        ]
+
+        for par, name in zip(tqdm.tqdm(variables), description_names):
+            varnames = ["mean_" + par, "d" + par + "dx", "d" + par + "dy"]
+            var_units = units(self.circle_ds[par].attrs.get("units", None))
+            descriptions = [
+                "Circle mean of " + name,
+                "Zonal gradient of " + name,
+                "Meridional gradient of " + name,
+            ]
+
+            results = self.fit2d_xr(
+                x=self.circle_ds.dx,
+                y=self.circle_ds.dy,
+                u=self.circle_ds[par],
+                sonde_dim="sonde_id",
+            )
+
+            assign_dict = {}
+            for varname, result, description in zip(varnames, results, descriptions):
+                assign_dict[varname] = (
+                    ["alt"],
+                    result.data,
+                )  # Assign data without attrs for now
+
+            # Assign the new variables to the dataset
+            self.circle_ds = self.circle_ds.assign(assign_dict)
+
+            # Now, add the attributes (description, units) after assignment
+            for varname, description in zip(varnames, descriptions):
+                self.circle_ds[varname].attrs["description"] = description
+
+                if "mean" in varname:
+                    result_units = "{:~}".format(var_units.units)
+                elif "dq" in varname:
+                    result_units = "kg / kg / m"
+                else:
+                    result_units = "{:~}".format((var_units / units.meters).units)
+
+                self.circle_ds[varname].attrs["units"] = result_units
```

Comment:
it would be good if the attribute assignment would happen in the `assign()` call and not have [] before the = in this code